ssh-agent.c

```diff
@@ -2368,8 +2367,27 @@ main(int ac, char **av)
 
 	parent_pid = getpid();
 
-	if (agentsocket == NULL) {
-		/* Create private directory for agent socket */
+	/* Has the socket been provided via socket activation? */
+	if (agentsocket == NULL && ac == 0 && (d_flag || D_flag) &&
+	    (pidstr = getenv("LISTEN_PID")) != NULL &&
+	    (fdstr = getenv("LISTEN_FDS")) != NULL) {
+		if (strcmp(fdstr, "1") != 0) {
+			fatal("unexpected LISTEN_FDS contents "
+			    "(want: \"1\" got\"%s\"", fdstr);
+		}
+		if (fcntl(3, F_GETFL) == -1)
+			fatal("LISTEN_FDS set but fd 3 unavailable");
+		pid = (int)strtonum(pidstr, 1, INT_MAX, &errstr);
+		if (errstr != NULL)
+			fatal("invalid LISTEN_PID: %s", errstr);
+		if (pid != getpid())
+			fatal("bad LISTEN_PID: %d vs pid %d", pid, getpid());
+		debug("using socket activation on fd=3");
+		sock = 3;
+	}
+
+	/* Otherwise, create private directory for agent socket */
+	if (sock == -1 && agentsocket == NULL) {
```

Comment:
Something is wrong with this predicate in the supervised case.  In the supervised case, `sock == 3` but `agentsocket == NULL`, which means that the `else` clause here will trigger.  but in the else clause, it tries to `strlcpy` from `agentsocket`.  This will result in a segfault.

Reply:
Sorry, I just pushed a fix for this. Can you give it another try?

Reply:
yes, it looks like it works in the supervised case now.  thanks for the fix!